trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/cfg/ProcedureMultiplier.java

```diff
@@ -86,182 +84,173 @@
  *
  *
  * @author Matthias Heizmann (heizmann@informatik.uni-freiburg.de)
+ * @author Frank Schüssele (schuessf@informatik.uni-freiburg.de)
  */
 public class ProcedureMultiplier {
-
-	public ProcedureMultiplier(final IUltimateServiceProvider services, final BasicIcfg<IcfgLocation> icfg,
-			final HashRelation<String, String> copyDirectives, final BlockEncodingBacktranslator backtranslator,
+	public static void duplicateProcedures(final IUltimateServiceProvider services, final BasicIcfg<IcfgLocation> icfg,
+			final List<ThreadInstance> instances, final BlockEncodingBacktranslator backtranslator,
 			final Map<IIcfgForkTransitionThreadCurrent<IcfgLocation>, List<ThreadInstance>> threadInstanceMap,
 			final List<IIcfgForkTransitionThreadCurrent<IcfgLocation>> forkCurrentThreads,
-			final List<IIcfgJoinTransitionThreadCurrent<IcfgLocation>> joinCurrentThreads) {
-		final IcfgEdgeFactory icfgEdgeFactory = icfg.getCfgSmtToolkit().getIcfgEdgeFactory();
-		final Map<String, List<ILocalProgramVar>> inParams = new HashMap<>(icfg.getCfgSmtToolkit().getInParams());
-		final Map<String, List<ILocalProgramVar>> outParams = new HashMap<>(icfg.getCfgSmtToolkit().getOutParams());
-		final Set<String> procedures = new HashSet<>(icfg.getCfgSmtToolkit().getProcedures());
-		final SmtFunctionsAndAxioms smtSymbols = icfg.getCfgSmtToolkit().getSmtFunctionsAndAxioms();
+			final List<IIcfgJoinTransitionThreadCurrent<IcfgLocation>> joinCurrentThreads, final Object lockOwner) {

```

Comment:
What's the reason for this `lockOwner` parameter?

Reply:
It is needed for the `ManagedScript`, before we could just use `this`, but this is no more possible since it is a static method.

Reply:
Hmm, ok. But wouldn't it still be better to use some available object for this?

* This is an implementation detail, the caller has no interest in which object is used.
* The caller might misunderstand the API, thinking they can pass an already-locked `ManagedScript` along with the lock owner. But this will fail.

Reply:
I think we often carry such a `lockOwner`-object. Are there any better ideas? Should we just use e.g. `ProcedureMultiplier.class` instead?

Reply:
My impression was that we only do this in cases where the ManagedScript is already locked, but I may be wrong. The class object sounds like a good idea. some methods e.g. in `TransformulaUtils` use other things, e.g. the `TransFormula`, but the class object sounds like a cleaner approach that will produce more helpful error messages.